[lldb] fix parallel module loading deadlock for Linux DYLD

zhyty opened the pull request:
Another attempt at resolving the deadlock issue @GeorgeHuyubo  discovered (his previous [attempt](https://github.com/llvm/llvm-project/pull/160225)). 

This change can be summarized as the following:
* Plumb through a boolean flag to force no preload in `GetOrCreateModules` all the way through to `LoadModuleAtAddress`.
* Parallelize `Module::PreloadSymbols` separately from `Target::GetOrCreateModule` and its caller `LoadModuleAtAddress` (this is what avoids the deadlock).

These changes roughly maintain the performance characteristics of the previous implementation of parallel module loading. Testing on targets with between 5000 and 14000 modules, I saw similar numbers as before, often more than 10% faster in the new implementation across multiple trials for these massive targets. I think it's because we have less lock contention with this approach.

# The deadlock

See [bt.txt](https://github.com/user-attachments/files/22524471/bt.txt) for a sample backtrace of LLDB when the deadlock occurs.

As @GeorgeHuyubo explains in his [PR](https://github.com/llvm/llvm-project/pull/160225), the deadlock occurs from an ABBA deadlock that happens when a thread context-switches out of `Module::PreloadSymbols`, goes into `Target::GetOrCreateModule` for another module, possibly entering this block:
```
      if (!module_sp) {
        // The platform is responsible for finding and caching an appropriate
        // module in the shared module cache.
        if (m_platform_sp) {
          error = m_platform_sp->GetSharedModule(
              module_spec, m_process_sp.get(), module_sp, &search_paths,
              &old_modules, &did_create_module);
        } else {
          error = Status::FromErrorString("no platform is currently set");
        }
      }
```
`Module::PreloadSymbols` holds a module-level mutex, and then `GetSharedModule` *attempts* to hold the mutex of the global shared `ModuleList`. So, this thread holds the module mutex, and waits on the global shared `ModuleList` mutex.

A competing thread may execute `Target::GetOrCreateModule`, enter the same block as above, grabbing the global shared `ModuleList` mutex. Then, in `ModuleList::GetSharedModule`, we eventually call `ModuleList::FindModules` which eventually waits for the `Module` mutex held by the first thread (via `Module::GetUUID`). Thus, we deadlock.

## Reproducing the deadlock

It might be worth noting that I've never been able to observe this deadlock issue during live debugging (e.g. launching or attaching to processes), however we were able to consistently reproduce this issue with coredumps when using the following settings:
```
(lldb) settings set target.parallel-module-load true
(lldb) settings set target.preload-symbols true
(lldb) settings set symbols.load-on-demand false
(lldb) target create --core /some/core/file/here
# deadlock happens
```

## How this change avoids this deadlock

This change avoids concurrent executions of `Module::PreloadSymbols` with `Target::GetOrCreateModule` by waiting until after the `Target::GetOrCreateModule` executions to run `Module::PreloadSymbols` in parallel. This avoids the ordering of holding a Module lock *then* the ModuleList lock, as `Target::GetOrCreateModule` executions maintain the ordering of the shared ModuleList lock first (from what I've read and tested).

## Why not read-write lock?

Some feedback in https://github.com/llvm/llvm-project/pull/160225 was to modify mutexes used in these components with read-write locks. This might be a good idea overall, but I don't think it would *easily* resolve this specific deadlock. `Module::PreloadSymbols` would probably need a write lock to Module, so even if we had a read lock in `Module::GetUUID` we would still contend. Maybe the `ModuleList` lock could be a read lock that converts to a write lock if it chooses to update the module, but it seems likely that some thread would try to update the shared module list and then the write lock would contend again. 

Perhaps with deeper architectural changes, we could fix this issue?

# Other attempts

One downside of this approach (and the former approach of parallel module loading) is that each DYLD would need to implement this pattern themselves. With @clayborg's help, I looked at a few other approaches:
* In `Target::GetOrCreateModule`, backgrounding the `Module::PreloadSymbols` call by adding it directly to the thread pool via `Debugger::GetThreadPool().async()`. This required adding a lock to `Module::SetLoadAddress` (probably should be one there already) since `ObjectFileELF::SetLoadAddress` is not thread-safe (updates sections). Unfortunately, during execution, this causes the preload symbols to run synchronously with `Target::GetOrCreateModule`, preventing us from truly parallelizing the execution.
* In `Module::PreloadSymbols`, backgrounding the `symtab` and `sym_file` `PreloadSymbols` calls individually, but similar issues as the above.
* Passing a callback function like https://github.com/swiftlang/llvm-project/pull/10746 instead of the boolean I use in this change. It's functionally the same change IMO, with some design tradeoffs:
  * Pro: the caller doesn't need to explicitly call `Module::PreloadSymbols` itself, and can instead call whatever function is passed into the callback.
  * Con: the caller needs to delay the execution of the callback such that it occurs after the `GetOrCreateModule` logic, otherwise we run into the same issue. I thought this would be trickier for the caller, requiring some kinda condition variable or otherwise storing the calls to execute afterwards. 

# Test Plan:
```
ninja check-lldb
```

Commit:
move preload out of GetOrCreateModules and parallelize

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: https://phabricator.intern.facebook.com/D86004440

Commit:
formatting fix

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

zhyty commented:
> I wonder if instead of a boolean we should pass down a function that schedules the module loading as suggested in #10746.
> 
> That gives more flexibility in that we could immediately schedule the preload to happen in parallel (as done in the other PR), or for your use case we could accumulate the preload callbacks into a vector and then schedule all the preload jobs to run in parallel once all the modules have been loaded.
> 
> Any thoughts @DmT021?

(Proper link would be https://github.com/swiftlang/llvm-project/pull/10746. #10746 in this repo refers to something unrelated)

zhyty commented:
> I wonder if instead of a boolean we should pass down a function that schedules the module loading as suggested in #10746.
> 
> That gives more flexibility in that we could immediately schedule the preload to happen in parallel (as done in the other PR), or for your use case we could accumulate the preload callbacks into a vector and then schedule all the preload jobs to run in parallel once all the modules have been loaded.
> 
> Any thoughts @DmT021?

I just finished writing a small blurb explaining why I went with this approach, so maybe that helps justify this approach over that one?

> That gives more flexibility in that we could immediately schedule the preload to happen in parallel (as done in the other PR)

Immediately scheduling the preload to happen in parallel shouldn't be thread-safe due to the ABBA deadlock. I don't think that flexibility is good, then, if it leads callers into this deadlock.

JDevlieghere commented:
Thanks for the backtrace and the clarification. I now also finally understand why you mentioned "context switching", which confused me as that's not something the LLVM ThreadPool supports. What's going on is that one of the worker threads is submitting new tasks (belonging to a different task group) and waiting on them. 

From https://llvm.org/doxygen/classllvm_1_1ThreadPoolInterface.html: 
> It is also possible for worker threads to submit new tasks and wait for them. Note that this may result in a deadlock in cases such as when a task (directly or indirectly) tries to wait for its own completion, or when all available threads are used up by tasks waiting for a task that has no thread left to run on (this includes waiting on the returned future). It should be generally safe to [wait()](https://llvm.org/doxygen/classllvm_1_1ThreadPoolInterface.html#a0ea46b29f97f4f535c563a68a38185a4) for a group as long as groups do not form a cycle.

However that last statement doesn't account for the situation where we're in. IIUC, we have a task group X and a task group Y. We create task group Y  from a task belonging to X, and when we call wait, the thread pool picks up the next task from X, not from Y. This is what you referred to as "context switching". It also explains why @GeorgeHuyubo's switch to a separate ThreadPool avoided the issue. 

With that information, I would argue that the actual mutexes involved here are irrelevant. Nothing can make this operation inherently safe. Even without any mutexes, with a sufficiently small group of threads, this may never complete. The only solution here is to not create a new task pool from a task. Your patch is doing exactly that by deferring the preload until after we're processing the modules in parallel. 

Alright, I've convinced myself that this is the right approach. Thanks for bearing with me. I'm now also convinced that we should do this unconditionally. The only reason we're not seeing this on Darwin is because we always use the accelerator tables and don't need to manually index. However, it inherently suffers from the same problem. Let's change this globally and we should be good everywhere.

Commit:
remove defer_module_preload bool flag, remove preload from
Target::GetOrCreateModule

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

zhyty committed:
Merge branch 'main' into parallel-module-load-fix-oss

JDevlieghere reviewed (approved):
Thank you, this LGTM modulo nits.

Commit:
address coding style issues

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Commit:
remove explicit preloads -- move preloading to Target::ModulesDidLoad

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Commit:
remove small erroneous formatting change in wasmdyld

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Commit:
tweak Target.h documentation

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

zhyty committed:
Merge branch 'main' into parallel-module-load-fix-oss

clayborg reviewed (approved):
LGTM. @JDevlieghere let us know if you like the `Target::ModulesDidLoad` doing the preloading and obeying the target settings.

JDevlieghere reviewed (approved):
Thanks for the ping. Doing this in `ModulesDidLoad` makes sense. LGTM.
